src/coreclr/src/debug/createdump/datatarget.cpp

```diff
@@ -149,8 +134,9 @@ DumpDataTarget::ReadVirtual(
     /* [in] */ ULONG32 size,
     /* [optional][out] */ ULONG32 *done)
 {
-    assert(m_fd != -1);
-    ssize_t read = pread64(m_fd, buffer, size, (off64_t)(ULONG_PTR)address);
+    iovec local{ buffer, size };
+    iovec remote{ (void*)(ULONG_PTR)address, size };
+    ssize_t read = process_vm_readv(m_pid, &local, 1, &remote, 1, 0);
```

Comment:
Also, are there any places where it would be beneficial to pass an `iovec` array rather than just 1 element at a time to `process_vm_readv()`?  (There could be a fallback method to go through each one and read the memory on OSes that do not offer this particular syscall or a similar feature.)

It might be a good idea to `#ifdef` this for Linux only, or check if the syscall exist during build-time.  For instance, the FreeBSD `procfs` supports `/proc/<PID>/mem` but does not have this syscall, so this breaks FreeBSD.  On systems without `process_vm_readv()` but with a compatible `procfs`, you might benefit from using `readv()` instead of `pread()` and have similar speedups if the suggestion above is possible.

Reply:
There may even be Linux distros that don't support process_vm_readv so a build-time check (like all the other HAVE_xxxx defines) should be added falling back to the proc/<PID>/mem code.

Reply:
| Distro        | Version | Kernel |
| ------------- | ------- | ------ |
| RHEL          | 7.0     | 3.10   |
| CentOS        | 7.0     | 3.10   |
| Oracle Linux  | 7.0     | 3.10   |
| Fedora        | 29      | 4.18   |
| Debian        | 9       | 4.9    |
| Ubuntu        | 16.04   | 4.4    |
| Linux Mint    | 18      | 4.4    |
| openSUSE      | 15      | 4.12   |
| SLES          | 12 SP2  | 4.4    |
| Alpline Linux | 3.8     | 4.14   |

> There may even be Linux distros that don't support process_vm_readv

All supported Linux distros meet the [minimum version requirement](https://github.com/dotnet/runtime/pull/420#issuecomment-560127405), so I assume it only needs to check for Linux.

Reply:
> For instance, the FreeBSD `procfs` supports `/proc/<PID>/mem`

The [`createdump` doc](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/xplat-minidump-generation.md#freebsdopenbsdnetbsd) says `/proc` doesn't work on FreeBSD. Can you confirm it? Thanks!

Reply:
> Also, are there any places where it would be beneficial to pass an iovec array rather than just 1 element at a time to process_vm_readv()?

There are some parallel read operations, but they are invoked on `ICLRDataTarget` instead of `DumpDataTarget`. It's not easy to extend that interface.

Reply:
> you might benefit from using `readv()` instead of `pread()` and have similar speedups if the suggestion above is possible.

`readv`/`preadv` simply wraps `read`/`pread` and therefore does not bring speedups.

Reply:
I wrote that createdump doc 2 or 3 years ago and at the time, it didn't look like FreeBSD supported /proc/<PID>/mem but now I'm not sure.